### DISCUSSION TITLE
Overloading default constructors for uvm_objects

### DISCUSSION THREAD

**Discussion by timothyscherer** [pyuvm/pyuvm#183] [General]
I noticed in the book, there are instances (such as with the `OpSeq` on page 315) where the `__init__` gets parameterized. This is certainly more convenient than having to assign the variables after creation, and much more pythonic. However, this is not compatible with the uvm_factory create() method for overriding by name or type.

Is this bad practice, or should we add `**kwargs` to the factory method?

**Comment by raysalemi** [comment]
Are you sure it doesn't work with the factory?

**Reply by raysalemi** [reply]
I don't believe in making all classes factory compatible. The class developer needs to decide,and if they want a class to be factory compatible, they need to use the `ConfigDB()` to share data with an object.

**Reply by timothyscherer** [reply]
Using the ConfigDB to configure something such as a sequence item is not very scalable, since the sequence is not a uvm_component, so it cannot be used as context for setting configuration. For example, the book uses the key "SEQR" with no context, and an empty instance name. If you ever had two of these sequences with unique sequencers, this approach is no longer effective.

The UVM examples I see create the item then assign its attributes in the lines following the creation, which seems to go against the goal of keeping pyuvm pythonic, since class attributes are getting created and assigned from outside the class.

Adding arguments to the factory create could work as follows, just allowing args and kwargs to be passed through the factory creation. Would this be acceptable, or too much of a deviation from UVM philosophy? I believe it is only like this because SystemVerilog does not allow overloading function arguments.
`s05_base_classes.py`
```py
# 5.3.5.1
    @classmethod
    def create(cls, name, *args, *kwargs):
        """
        :return: new object from factory
        """
        new_obj = uvm_factory().create_object_by_type(cls, name=name, *args, **kwargs)
        return new_obj
```

`s08_factory_classes.py`
```py
def create_object_by_type(self, requested_type, parent_inst_path="",
                              name="", *args, **kwargs):
        """
        8.3.1.5 Creation
        :param requested_type: The type that we request but that can be
        overridden
        :param parent_inst_path: The get_full_name path of the parent
        :param name: The name of the instance requested_type("name")
        :return: Type that is child of uvm_object.
        If the type is is not in the factory we raise UVMFactoryError
        """
        new_type = self.__find_override(requested_type, parent_inst_path, name)
        if new_type is None:
            raise error_classes.UVMFactoryError(
                f"{requested_type} not in uvm_factory()")
        return new_type(name, *args, **kwargs)
```

**Reply by raysalemi** [reply]
It is too much of a deviation from the 1800.2 spec.
